File: lib/buttons/button-checkbox.dart
import "package:angular2/angular2.dart";

@Directive(
    selector: "n2s-btn-checkbox",
    inputs: const ["trueValue", "falseValue"],
    host: const { "(click)" : "onClick()", "[class.active]" : "state"}
)
class ButtonCheckbox extends DefaultValueAccessor implements OnInit {
  ButtonCheckbox(this.ngModel, Renderer renderer, ElementRef elementRef) : super(renderer, elementRef) {
    ngModel.valueAccessor = this;
  }

  NgModel ngModel;

  dynamic trueValue = true;

  dynamic falseValue = false;

  var _value;

  bool state = false;


  @override
  ngOnInit() {
    state = trueValue == _value;
  }

  writeValue(value) {
    _value = value;
    state = trueValue == _value;
    super.writeValue(_value);
  }

  toggle(bool _state) {
    state = _state;
    _value = state ? trueValue : falseValue;
    ngModel.viewToModelUpdate(_value);
  }

  onClick() {
    toggle(!state);
  }
}

File: lib/buttons/button-checkbox.ts
/// <reference path="../../tsd.d.ts" />
import {
  Directive,
  DefaultValueAccessor,
  Self, NgModel, Renderer, ElementRef,
  OnInit
} from 'angular2/angular2';


@Directive({
  selector: '[btn-checkbox][ng-model]',
  properties: ['btnCheckboxTrue', 'btnCheckboxFalse'],
  host: {
    '(click)': 'onClick()',
    '[class.active]': 'state'
  }
})
export class ButtonCheckbox extends DefaultValueAccessor implements OnInit {
  private btnCheckboxTrue:any;
  private btnCheckboxFalse:any;
  private value:any;
  private state:boolean = false;

  constructor(@Self() cd:NgModel, renderer:Renderer, elementRef:ElementRef) {
    super(cd, renderer, elementRef);
  }

  ngOnInit() {
    this.toggle(this.trueValue === this.value);
  }

  private get trueValue() {
    return typeof this.btnCheckboxTrue !== 'undefined' ? this.btnCheckboxTrue : true;
  }

  private get falseValue() {
    return typeof this.btnCheckboxFalse !== 'undefined' ? this.btnCheckboxFalse : false;
  }

  toggle(state:boolean) {
    this.state = state;
    this.value = this.state ? this.trueValue : this.falseValue;
  }

  // model -> view
  writeValue(value:any) {
    this.state = this.trueValue === value;
    this.value = value;
  }

  // view -> model
  onClick() {
    this.toggle(!this.state);
    this.cd.viewToModelUpdate(this.value);
  }
}
